lib/Transforms/LiquidFixpointBuilder/VariablesEnvironmentImmutable.cpp

```diff
@@ -0,0 +1,577 @@
+  }
+  
+  ResultType VariablesEnvironmentImmutable::getCommonVariables(std::vector<std::string> previousBlocks, std::set<std::string>& commonVariables)
+  {
+    std::vector<std::string> previousFinishedBlocks;
+    std::vector<std::string> previousUnfinishedBlocks;
+
+    for (auto& previousBlock: previousBlocks)
+    {
+      RefinementUtils::Contains(finishedBlocks, previousBlock) ?
+	previousFinishedBlocks.push_back(previousBlock) :
+	previousUnfinishedBlocks.push_back(previousBlock);
+    }
+
+    // In a CFG, it is possible that previous blocks aren't _all_ processed (loops)
+    // However, for immutable constructs, this shouldn't happen. Raise an Error, if it does.
+    if (previousUnfinishedBlocks.size() > 0)
+    {
+      return ResultType::Error("There exist: "s + std::to_string(previousUnfinishedBlocks.size()) + " unfinished Block that were appended without processing. This is an error in an immutable environment."s);
+    }
+
+    bool first = true;
+
+    for (auto& previousBlock : previousBlocks)
+    {
+      if (first)
+      {
+	commonVariables = RefinementUtils::GetKeysSet(variablesMappingsPerBlock[previousBlock]);
+	first = false;
+      }
+      else
+      {
+	std::set<std::string> currBlockVars = RefinementUtils::GetKeysSet(variablesMappingsPerBlock[previousBlock]);
+	commonVariables = RefinementUtils::SetIntersection(commonVariables, currBlockVars);
+      }
+    }
+    return ResultType::Success();
+  }
+
+  //@TODO::juspreet - Continue ripping and verifying.
+  ResultType VariablesEnvironmentImmutable::getBlockBindersForPhiNode(const std::string& previousBlock, const std::string& blockName, const std::string& mappedVariableName, std::vector<std::string>& binders)
+  {
+    binders = getBlockBinders(previousBlock);
+
+    // Add the transition Edge Type to the Environment
+    std::string transitionGuardName = "__transition__"s + previousBlock + "__"s + currentBlockName;
+    binders.emplace_back(transitionGuardName);
+
+    std::string predecessorEntryGuard;
+    {
+      auto getPredEntryRes = getBlockGuard(previousBlock, predecessorEntryGuard);
+      if (!getPredEntryRes.Succeeded) { return getPredEntryRes; }
+    }
+
+    std::string phiNodeTransitionGuard = predecessorEntryGuard + " && "s + transitionGuardName;
+    std::string phiNodeGuardName = "__phi__"s + mappedVariableName + "_"s + previousBlock;
+
+    {
+      auto createBinderRes = constraintBuilder.CreateBinderWithConstraints(phiNodeGuardName, FixpointType::GetBoolType(), { phiNodeTransitionGuard });
+      if (!createBinderRes.Succeeded) { return createBinderRes; }
+    }
+
+    binders.emplace_back(phiNodeGuardName);
+    return ResultType::Success();
+  }
+
+  ResultType VariablesEnvironmentImmutable::createPhiNodeWithoutCreatedBinders(const std::string& variable, const std::string& mappedVariableName, const FixpointType& type, const std::vector<std::string>& sourceVariableNames, const std::vector<std::string>& previousBlocks)
```

Comment:
createPhiNodeAssumingNoFutureBinders

Reply:
I revered this, as this was wrong. We were _not_ using `createPhiNodeAssumingNoFutureBinders` and were actually using `createPhiNodeWithoutCreatedBinders`. The functionality wasn't different, as there is a statement check for undeclared variables that we will never go into (save for the sake of `for` loops).